### PR TITLE
Regional Prompting (Node only)

### PR DESCRIPTION
**Pull request by RyanJDick**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [x] Yes
- [ ] No, because:

      
## Have you updated all relevant documentation?
- [ ] Yes
- [x] No. Not yet.

## Description

This branch makes the following changes to support regional prompting:
- Adds a "Create Rectangle Mask" invocation for easily creating masks with nodes.
- Adds a `mask` input to the compel prompt invocations (for both SD and SDXL)
- Changes the "Denoise Latents" invocation to accept a collection of positive and negative text conditioning inputs.
- Updates the UNet attention modules to apply regional masking in the cross-attention layers.

Here's a sample workflow using the new nodes:

image

## Note about Memory Usage:

When IP-Adapter and/or regional prompting are used, we use a custom attention processor. This attention processor does not currently support xformers or sliced attention, so will use more memory than standard attention when those options are enabled.

The custom attention processor currently uses [torch.scaled_dot_product_attention()](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html).

If there is enough demand, we _could_ add support for xformers and sliced implementations. But, it probably makes sense to re-think our attention configuration strategy in the context of the latest improvements to torch (which supports low-memory and flash-attention modes).

### Completed Tests
Basic functionnality:
- [x] CUDA, SDXL, Standard
- [x] CUDA, SD, Standard
- [x] CUDA, SDXL, Sequential
- [x] CUDA, SD, Sequential
- [x] CPU, SDXL, Standard

Speed regression tests (run on an RTX4090):
- [x] No regression in speed for standard global prompting
- [x] The speed cost of using multiple prompts in DenoiseLatents is small. With 3 prompts, the denoise latents time increases from ~9.6s to ~10.2s. The more significant cost is having to run the text encoder N times.
- [x] Speed of IP-adapter is unchanged.

Compatibility:
- [x] Regional negative prompts
- [x] No global prompt, regions do not 100% cover the image.
- [x] Compatible with IP-Adapter
- [x] Compatible with T2I-Adapter
- [x] Compatible with ControlNet

### Remaining tests

- [ ] ROCm
- [ ] mps

## Added/updated tests?

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by lstein — COMMENTED** [review]
Impressive work. A couple of questions.

- We now have several masks being used at multiple places, including the bitmap inpainting mask, the boolean mask used in regional prompting, and the segment anything mask used in #5829 . Are these compatible with each other? For example, can I use a segment anything mask as a regional prompting mask?
- What is the use case for a rectangular prompting mask?

**Comment by hipsterusername** [comment]
> Impressive work. A couple of questions.
> 
>     * We now have several masks being used at multiple places, including the bitmap inpainting mask, the boolean mask used in regional prompting, and the segment anything mask used in [wip: Segment Anything #5829](https://github.com/invoke-ai/InvokeAI/pull/5829) . Are these compatible with each other? For example, can I use a segment anything mask as a regional prompting mask?
> 
>     * What is the use case for a rectangular prompting mask?

Re: Masks - That is the intent. They should just be Alpha channel-only images, interchangeable in different contexts. The system does not currently seem to handle "Mask" & "Image" as distinct types, but more as useful labels of the same type of object.

The Rectangular Mask is primarily for creating a Mask from a workflow without importing an existing image. It's obviously not the _ideal_ UX - but we currently lack a better interface in the Workflow editor if the user doesn't want to leave it to create the mask. The nature of the regional prompting ultimately doesn't require an exact shape, so the rectangle is "good enough" for many use-cases.

**Comment by RyanJDick** [comment]
From a quick skim, all of our new mask formats are likely not compatible. I'll look into fixing that.

**Comment by RyanJDick** [comment]
I started a thread here to discuss consolidating the mask types: https://discord.com/channels/1020123559063990373/1225113849489915925/1225113851121504416

**Comment by psychedelicious** [comment]
I've got a branch with a WIP UI (not hooked up to any graphs). It outputs an segmented mask image like this, which I'm using for testing:
canvas

I've added a couple nodes to this branch:
- `ExtractMasksAndPromptsInvocation`: Given a segmented mask image and list of prompt-color pairs, derives a mask for each prompt-color pair, outputting a list of prompt-mask pairs. There's a default value for the list of of pairs that matches the above test image.
- `SplitMaskPromptPair`: Splits a single prompt-mask pair into a prompt and `MaskField`.

I haven't been able to generate with regional prompts, though.

When I recreate @RyanJDick 's example workflow, I get this error:
```
[2024-04-08 18:00:51,743]::[InvokeAI]::ERROR --> Error while invoking session e714ce61-bc0a-493b-ab9d-f6888d06bea7, invocation 8386e799-d3ca-4456-a1f1-24d6e17ba051 (denoise_latents):
Linear.forward() takes 2 positional arguments but 3 were given
[2024-04-08 18:00:51,743]::[InvokeAI]::ERROR --> Traceback (most recent call last):
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/services/session_processor/session_processor_default.py", line 179, in _process
    outputs = self._invocation.invoke_internal(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/baseinvocation.py", line 281, in invoke_internal
    output = self.invoke(context)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/latent.py", line 937, in invoke
    result_latents = pipeline.latents_from_embeddings(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/stable_diffusion/diffusers_pipeline.py", line 359, in latents_from_embeddings
    latents = self.generate_latents_from_embeddings(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/stable_diffusion/diffusers_pipeline.py", line 452, in generate_latents_from_embeddings
    step_output = self.step(
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/stable_diffusion/diffusers_pipeline.py", line 573, in step
    uc_noise_pred, c_noise_pred = self.invokeai_diffuser.do_unet_step(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/stable_diffusion/diffusion/shared_invokeai_diffusion.py", line 241, in do_unet_step
    ) = self._apply_standard_conditioning(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/stable_diffusion/diffusion/shared_invokeai_diffusion.py", line 376, in _apply_standard_conditioning
    both_results = self.model_forward_callback(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/stable_diffusion/diffusers_pipeline.py", line 646, in _unet_forward
    return self.unet(
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/diffusers/models/unets/unet_2d_condition.py", line 1216, in forward
    sample, res_samples = downsample_block(
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/diffusers/models/unets/unet_2d_blocks.py", line 1279, in forward
    hidden_states = attn(
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/diffusers/models/transformers/transformer_2d.py", line 397, in forward
    hidden_states = block(
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/diffusers/models/attention.py", line 329, in forward
    attn_output = self.attn1(
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/diffusers/models/attention_processor.py", line 522, in forward
    return self.processor(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/stable_diffusion/diffusion/custom_atttention.py", line 115, in __call__
    query = attn.to_q(hidden_states, *args)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
TypeError: Linear.forward() takes 2 positional arguments but 3 were given
```

I adapted the workflow to SD1.5 and get the same error:
Screenshot 2024-04-08 at 6 25 41 pm

[Rectangle Regional Prompts - SD1.5.json](https://github.com/invoke-ai/InvokeAI/files/14902936/Rectangle.Regional.Prompts.-.SD1.5.json)

```
[2024-04-08 18:24:55,244]::[InvokeAI]::ERROR --> Error while invoking session bc2f25fa-59bc-4fed-b1b0-7ed133382ef8, invocation 3add961c-2fe6-4db8-96b2-53364d02d064 (denoise_latents):
Linear.forward() takes 2 positional arguments but 3 were given
[2024-04-08 18:24:55,244]::[InvokeAI]::ERROR --> Traceback (most recent call last):
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/services/session_processor/session_processor_default.py", line 179, in _process
    outputs = self._invocation.invoke_internal(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/baseinvocation.py", line 281, in invoke_internal
    output = self.invoke(context)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/latent.py", line 937, in invoke
    result_latents = pipeline.latents_from_embeddings(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/stable_diffusion/diffusers_pipeline.py", line 359, in latents_from_embeddings
    latents = self.generate_latents_from_embeddings(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/stable_diffusion/diffusers_pipeline.py", line 452, in generate_latents_from_embeddings
    step_output = self.step(
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/stable_diffusion/diffusers_pipeline.py", line 573, in step
    uc_noise_pred, c_noise_pred = self.invokeai_diffuser.do_unet_step(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/stable_diffusion/diffusion/shared_invokeai_diffusion.py", line 241, in do_unet_step
    ) = self._apply_standard_conditioning(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/stable_diffusion/diffusion/shared_invokeai_diffusion.py", line 376, in _apply_standard_conditioning
    both_results = self.model_forward_callback(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/stable_diffusion/diffusers_pipeline.py", line 646, in _unet_forward
    return self.unet(
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/diffusers/models/unets/unet_2d_condition.py", line 1216, in forward
    sample, res_samples = downsample_block(
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/diffusers/models/unets/unet_2d_blocks.py", line 1279, in forward
    hidden_states = attn(
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/diffusers/models/transformers/transformer_2d.py", line 397, in forward
    hidden_states = block(
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/diffusers/models/attention.py", line 329, in forward
    attn_output = self.attn1(
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/diffusers/models/attention_processor.py", line 522, in forward
    return self.processor(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/stable_diffusion/diffusion/custom_atttention.py", line 115, in __call__
    query = attn.to_q(hidden_states, *args)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
TypeError: Linear.forward() takes 2 positional arguments but 3 were given
```

I tried with `attention_type: torch-sdp` and xformers, same error on both. I had merged `main` into this branch, but switching to the last pre-merge commit and re-installing dependencies doesn't change it.

---

I get a different error if I use the new nodes I created. Here's the workflow:

[Arbitrary Regional Prompts.json](https://github.com/invoke-ai/InvokeAI/files/14902823/Arbitrary.Regional.Prompts.json)

image

And the error:
```
[2024-04-08 18:16:47,497]::[InvokeAI]::ERROR --> Error while invoking session 76149546-6732-4097-ad5a-365914372f4a, invocation eede7508-9af7-41d4-89bf-0a0668eef7c1 (denoise_latents):

[2024-04-08 18:16:47,497]::[InvokeAI]::ERROR --> Traceback (most recent call last):
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/services/session_processor/session_processor_default.py", line 179, in _process
    outputs = self._invocation.invoke_internal(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/baseinvocation.py", line 281, in invoke_internal
    output = self.invoke(context)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/latent.py", line 909, in invoke
    conditioning_data = self.get_conditioning_data(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/latent.py", line 517, in get_conditioning_data
    cond_text_embedding, cond_regions = self._concat_regional_text_embeddings(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/latent.py", line 472, in _concat_regional_text_embeddings
    regions = TextConditioningRegions(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/stable_diffusion/diffusion/conditioning_data.py", line 86, in __init__
    assert self.masks.shape[1] == len(self.ranges)
AssertionError
```

I think I need a "global" positive prompt, so I added that and got a different error:

[Arbitrary Regional Prompts with global prompt.json](https://github.com/invoke-ai/InvokeAI/files/14902965/Arbitrary.Regional.Prompts.with.global.prompt.json)

image

```
[2024-04-08 18:28:11,117]::[InvokeAI]::ERROR --> Error while invoking session af3ddd99-2f63-49c0-8b37-c6cd5bcf5c15, invocation 7afcbe9e-8d12-42b6-9727-6989651fbb3b (denoise_latents):
Tensors must have same number of dimensions: got 3 and 4
[2024-04-08 18:28:11,117]::[InvokeAI]::ERROR --> Traceback (most recent call last):
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/services/session_processor/session_processor_default.py", line 179, in _process
    outputs = self._invocation.invoke_internal(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/baseinvocation.py", line 281, in invoke_internal
    output = self.invoke(context)
  File "/home/bat/.invokeai/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/latent.py", line 909, in invoke
    conditioning_data = self.get_conditioning_data(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/latent.py", line 517, in get_conditioning_data
    cond_text_embedding, cond_regions = self._concat_regional_text_embeddings(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/latent.py", line 473, in _concat_regional_text_embeddings
    masks=torch.cat(processed_masks, dim=1),
RuntimeError: Tensors must have same number of dimensions: got 3 and 4
```

I suspect I'm not creating the mask tensor in the correct shape in my node.

**Comment by psychedelicious** [comment]
Here's the very rough regional prompt UI (no feedback needed at this point, it's just functional enough for testing):

https://github.com/invoke-ai/InvokeAI/assets/4822129/76ca0e48-c396-4182-b839-eca89e91b3e3

**Comment by RyanJDick** [comment]
@psychedelicious I have addressed your first error (`TypeError: Linear.forward() takes 2 positional arguments but 3 were given`) in [98900a7](https://github.com/invoke-ai/InvokeAI/pull/5916/commits/98900a7ff1de2dea52c1742d0d87057096e77240). See that commit's message for more info. I wasn't hitting it, because I had peft installed, which was triggering slightly different behaviour.

**Comment by RyanJDick** [comment]
@psychedelicious The errors coming from your new nodes are addressed in [826f3d6](https://github.com/invoke-ai/InvokeAI/pull/5916/commits/826f3d625a61ecaeef59864aabb61af40ab5338a)

**Review by psychedelicious — APPROVED** [review]
Confirmed errors are resolved.

SD1.5 is pretty iffy in terms of putting the prompted things in the specified regions, but SDXL works great.

I've reverted my added node. It was built on incorrect assumptions about how regional prompting worked. I'll add a different node for arbitrary mask images in the PR that adds the UI.

File: invokeai/app/invocations/latent.py [hunk 1]
```diff
@@ -655,7 +813,15 @@ def init_scheduler(
...
         num_inference_steps = len(timesteps) // scheduler.order
 
-        return num_inference_steps, timesteps, init_timestep
+        scheduler_step_kwargs = {}
+        scheduler_step_signature = inspect.signature(scheduler.step)
+        if "generator" in scheduler_step_signature.parameters:
+            # At some point, someone decided that schedulers that accept a generator should use the original seed with
+            # all bits flipped. I don't know the original rationale for this, but now we must keep it like this for
+            # reproducibility.
+            scheduler_step_kwargs = {"generator": torch.Generator(device=device).manual_seed(seed ^ 0xFFFFFFFF)}
```
Comment (keturn) [inline invokeai/app/invocations/latent.py:822]: They needed an explicitly seeded generator passed for reproducibility.

Nobody wanted to add a new `stochastic_scheduler_seed` or `generator` field to DenoiseLatentsInvocation, but they figured out how to get a seed from one of its input LatentFields.

However, I feared re-using the same `seed` as had already been used to generate that input would result in the same tensor, which is not the random behavior those schedulers _expect_ to be doing.

So the compromise was, when you have an operation that needs a seed but the only seed defined has already been used earlier in the graph, you do some reproducible transformation like this to get a new seed for your new operation.

FWIW, I recommend factoring this `seed  ^ 0xFFFFFFFF` up out of this method to the place that calls `get_scheduler` and `init_scheduler`. That makes this method simpler: the seed passed to `init_scheduler` is, in fact, the seed used for them; and the quirky logic about deriving the `scheduler_seed` goes with the code that's trying to pull a `seed` out of the InvocationContext's LatentFields.

**Pull request merged**